verification/scratch/mwe2/mwe2.gobra

```diff
@@ -0,0 +1,132 @@
+pred token(ghost p Place)
+
+pred RecvPerm(ghost p Place)
+
+// Returns the next place after calling `Recv` from `p`.
+// Avoids using an existential quantifier.
+ghost
+requires RecvPerm(p)
+decreases
+pure func Recv_T(ghost p Place) Place
+
+// Used to refer to the received message.
+ghost
+requires RecvPerm(p)
+decreases
+pure func Recv_R(ghost p Place) int
+
+// This is how the state is/should be updated after receiving message `m`.
+decreases
+pure func Recv_S(s state, m int) state {
+    return state { key: s.key, lastMsg1: m, lastMsg2: s.lastMsg1, lastMsg3: s.lastMsg2 }
+}
+
+requires token(p) && RecvPerm(p)
+ensures token(old(Recv_T(p)))
+ensures m == old(Recv_R(p))
+ensures low(m) 
+func Recv(ghost p Place) (m int)
+
+pred SendPerm(ghost p Place, t int)
+
+ghost
+requires SendPerm(p, t)
+decreases
+pure func Send_T(ghost p Place, t int) Place
+
+requires token(p) && SendPerm(p, t)
+requires low(t)
+ensures token(old(Send_T(p, t)))
+func Send(ghost p Place, t int)
+
+pred DeclassifyPerm(ghost p Place, tag int, t int)
+
+ghost
+requires DeclassifyPerm(p, tag, t)
+decreases
+pure func Declassify_T(ghost p Place, tag int, t int) Place
+
+// In order to make permitted declassifications deterministic in low data,
+// we add low parameter `p` to "tag" declassifications.
+// (Necessary as otherwise IOD spec Protocol2 would not be well-formed.)
+ghost
+requires token(p) && DeclassifyPerm(p, tag, t)
+requires low(tag)
+ensures token(old(Declassify_T(p, tag, t)))
+ensures low(t)
+decreases
+func Declassify(ghost p Place, tag int, t int)
+
+// "Linear" protocol.
+pred Protocol1(ghost p Place, key int) {
+    // 1. Receive a message.
+    RecvPerm(p) && 
+    // 2. Compute MAC tag and declassify it.
+    DeclassifyPerm(Recv_T(p), Recv_R(p), MAC(key, Recv_R(p))) &&
+    // 3. Send MAC tag over network.
+    SendPerm(Declassify_T(Recv_T(p), Recv_R(p), MAC(key, Recv_R(p))), MAC(key, Recv_R(p))) && 
+    // 4. Restart.
+    Protocol1(Send_T(Declassify_T(Recv_T(p), Recv_R(p), MAC(key, Recv_R(p))), MAC(key, Recv_R(p))), key)
+}
+
+type state struct {
+    key int // the private key
+    lastMsg1 int // 1st most recently received message
+    lastMsg2 int // 2nd 
+    lastMsg3 int // 3rd 
+}
+
+pred Protocol2(ghost p Place, s state) {
+    // Receive a message at any time.
+    RecvPerm(p) && Protocol2(Recv_T(p), Recv_S(s, Recv_R(p))) &&
+    // NOTE: at the moment we can declassify things before receiving anything
+    // Declassify and send either the most or the 2nd most recently received message.
+    DeclassifyPerm(p, s.lastMsg1, MAC(s.key, s.lastMsg1)) && Protocol2(Declassify_T(p, s.lastMsg1, MAC(s.key, s.lastMsg1)), s) &&
+    DeclassifyPerm(p, s.lastMsg2, MAC(s.key, s.lastMsg2)) && Protocol2(Declassify_T(p, s.lastMsg2, MAC(s.key, s.lastMsg2)), s) &&
+    SendPerm(p, MAC(s.key, s.lastMsg1)) && Protocol2(Send_T(p, MAC(s.key, s.lastMsg1)), s) &&
+    SendPerm(p, MAC(s.key, s.lastMsg2)) && Protocol2(Send_T(p, MAC(s.key, s.lastMsg2)), s)
+}
```

Comment:
does this protocol show sth that the other did not?

Reply:
I constructed `Protocol2` to show a protocol ...
1. ... which allows declassifying different values at any point, and thus needs the additional parameter `p` for declassification (to make the IOD spec deterministic and the low data, *which includes* `p`).
2. ... where one is allowed by the IOD spec to send something before declassifying, which is only "caught" (in case it was high data) by the classification spec / pre- and postconditions of the I/O functions (in the other protocol, one had to declassify before sending just to conform to the IOD spec). This is supposed to exemplify that SIF goes beyond simply following the IOD spec.
3. ... which contains more complex, mutable state, and how to work with it.